### ISSUE TITLE
TFTRT and `tf.keras.application.ResNet50` INT8 and FP32 not working

### ISSUE THREAD

**Issue by benjamintanweihao** [tensorflow/tensorflow#24032]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): **Yes**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
**Ubuntu 18.04**
- TensorFlow installed from (source or binary): **binary**
- TensorFlow version (use command below): **tf-nightly (1.13.0-dev20181127).**
- Python version: **3.5**
- Bazel version (if compiling from source): **N/A**
- GCC/Compiler version (if compiling from source):  **N/A**
 CUDA/cuDNN version: **9.0 / 7.3**
- GPU model and memory: **GTX 1080 Ti (11GB)**

**Describe the current behavior**

Conversion to FP32 using TensorRT fails on `tf.keras.applications.InceptionResNetV2`, `tf.keras.applications.InceptionV3` with:

```
Engine my_trt_op_0 creation for segment 0, composed of 2049 nodes failed:
Invalid argument: Validation failed for TensorRTInputPH_0 and input slot 0:
Input tensor with shape [?,?,?,3] has an unknown non-batch dimemension at dim 1.
```

With `tf.keras.applications.VGG19`, conversion to FP32 fails with:

```
python: helpers.cpp:56: nvinfer1::DimsCHW nvinfer1::getCHW(const nvinfer1::Dims&): Assertion `d.nbDims >= 3' failed.
``` 

For `tf.keras.applications.ResNet50`, conversion to FP32 succeed but with INT8, the error is similar to `VGG19` with FP32:

```
python: helpers.cpp:56: nvinfer1::DimsCHW nvinfer1::getCHW(const nvinfer1::Dims&): Assertion `d.nbDims >= 3' failed.```
```

**Describe the expected behavior**

Given that it is possible to convert TensorFlow models such as Inception to TensorRT as evidenced by https://github.com/NVIDIA-AI-IOT/tf_trt_models, the conversion should similarly work for models in `tf.keras.applications.*`.

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.

```python
import os
os.environ['CUDA_VISIBLE_DEVICES'] = '0'
import tensorflow as tf
import tensorflow.contrib.tensorrt as trt
import utils
import time

from imutils import paths
from tensorflow.python.client import session as csess
from tensorflow.python.framework import importer as importer
from tensorflow.python.framework import ops as ops
from tensorflow.core.protobuf import config_pb2 as cpb2
from tensorflow.python.platform import gfile


print(tf.__version__)

gpu_memory = 10000000000
trt_frac = 0.5

# Create a session first...
sess = tf.Session(config=tf.ConfigProto(
    gpu_options=tf.GPUOptions(per_process_gpu_memory_fraction=trt_frac)))

tf.keras.backend.set_session(sess)

input_tensor = tf.placeholder(tf.float32, shape=(None, 224, 224, 3), name='image_tensor')

tf.keras.backend.set_learning_phase(0)

# NOTE: ResNet50 doesn't work with INT8
# python: helpers.cpp:56: nvinfer1::DimsCHW nvinfer1::getCHW(const nvinfer1::Dims&): Assertion `d.nbDims >= 3' failed.
model = tf.keras.applications.ResNet50(weights='imagenet')

# NOTE: MobileNetV2 doesn't work with INT8
# python: helpers.cpp:56: nvinfer1::DimsCHW nvinfer1::getCHW(const nvinfer1::Dims&): Assertion `d.nbDims >= 3' failed.
# model = tf.keras.applications.MobileNetV2(weights='imagenet')

# NOTE: InceptionResNetV2 and InceptionV3 doesn't work:
# Engine my_trt_op_0 creation for segment 0, composed of 2049 nodes failed:
# Invalid argument: Validation failed for TensorRTInputPH_0 and input slot 0:
# Input tensor with shape [?,?,?,3] has an unknown non-batch dimemension at dim 1.
# model = tf.keras.applications.InceptionResNetV2(weights='imagenet')
# model = tf.keras.applications.InceptionV3(weights='imagenet')

# NOTE: VGG19 doesn't work
# python: helpers.cpp:56: nvinfer1::DimsCHW nvinfer1::getCHW(const nvinfer1::Dims&): Assertion `d.nbDims >= 3' failed.
# model = tf.keras.applications.VGG19(weights='imagenet')

model_input = model.input.name.strip(':0')
model_output = model.output.name.strip(':0')

print(model_input, model_output)

graph = tf.get_default_graph().as_graph_def()

# freeze graph and remove nodes used for training
frozen_graph = tf.graph_util.convert_variables_to_constants(sess, graph, [model_output])
frozen_graph = tf.graph_util.remove_training_nodes(frozen_graph)

nodes = [node.name for node in frozen_graph.node]
print(nodes)

# Make this the default graph now
tf.import_graph_def(frozen_graph, name='')

max_workspace_size_bytes = int(trt_frac * gpu_memory)

precision_mode = 'INT8' # FP32 | INT8

trt_graph = trt.create_inference_graph(
    input_graph_def=frozen_graph,
    outputs=[model_output],
    max_batch_size=1,
    max_workspace_size_bytes=max_workspace_size_bytes,
    precision_mode=precision_mode,
    minimum_segment_size=3,
    maximum_cached_engines=3,
    is_dynamic_op=False
)

trt_engine_opts = len([1 for n in trt_graph.node if str(n.op) == 'TRTEngineOp'])
print('TRT Engine Op: {}'.format(trt_engine_opts))
assert trt_engine_opts > 0, 'No TRT Engine Ops!'

if precision_mode == 'INT8':
    with gfile.GFile("frozen/frozen_graph_INT8CALIB.pb", 'wb') as f:
        f.write(trt_graph.SerializeToString())
else:
    with gfile.GFile("frozen/frozen_graph_FP32.pb", 'wb') as f:
        f.write(trt_graph.SerializeToString())


if precision_mode == 'INT8':
    gpu_options = cpb2.GPUOptions(per_process_gpu_memory_fraction=trt_frac)
    sess_config = cpb2.ConfigProto(gpu_options=gpu_options)

    ops.reset_default_graph()
    g1 = ops.Graph()
    with g1.as_default():
        inp, out = importer.import_graph_def(
            graph_def=trt_graph,
            return_elements=[model_input, model_output]
        )
        inp = inp.outputs[0]
        out = out.outputs[0]
    with csess.Session(
            config=sess_config, graph=g1) as sess:
        for file_name in list(paths.list_images('samples'))[0:10]:
            input = utils.read_tensor_from_image_file(file_name,
                                                      input_height=224,
                                                      input_width=224)
            start_time = time.time()
            val = sess.run(out, {inp: input})
            stop_time = time.time()
            print('Inference time is {}'.format(stop_time - start_time))
            print('Running {}'.format(file_name))

    trt_graph = trt.calib_graph_to_infer_graph(trt_graph)

    with gfile.GFile("frozen/frozen_graph_INT8.pb", 'wb') as f:
        f.write(trt_graph.SerializeToString())

    ops.reset_default_graph()
    g2 = ops.Graph()
    with g2.as_default():
        # Run the inference

        inp, out = importer.import_graph_def(
            graph_def=trt_graph,
            return_elements=[model_input, model_output]
        )

        inp = inp.outputs[0]
        out = out.outputs[0]

    with csess.Session(
            config=sess_config, graph=g2) as sess:
        for file_name in list(paths.list_images('samples'))[0:10]:
            input = utils.read_tensor_from_image_file(file_name,
                                                      input_height=224,
                                                      input_width=224)
            start_time = time.time()
            val = sess.run(out, {inp: input})
            stop_time = time.time()
            print('Inference time is {}'.format(stop_time - start_time))
            print('Running {}'.format(file_name))
```

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

I have also tried messing around with every combination of `max_batch_size`, `minimum_segment_size`, `maximum_cached_engines` but no joy.

**Comment by trevor-m** [comment]
Hi, thanks for reporting the issue. 

* `Input tensor with shape [?,?,?,3] has an unknown non-batch dimemension at dim 1.`
Could you try with is_dynamic_op=True for create_inference_graph? TF-TRT requires all shapes to be fully defined unless in dynamic mode. You could also have the shapes be explicitly defined by doing the following:
```
input_tensor = Input(shape=(224, 224, 3))  # this assumes K.image_data_format() == 'channels_last'
model = InceptionV3(input_tensor=input_tensor)
```

*  `Assertion `d.nbDims >= 3' failed.`
Which version of TensorRT are you using? The message you are seeing is a bug in TRT 4.0 with INT8, which was fixed in TRT 5.0.

**Comment by samikama** [comment]
Closing this. Please reopen if @trevor-m 's suggestions don't work.
